ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,215 @@
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.util.KsqlConfig;
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.kafka.streams.KafkaStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Monitor and manages the restart of persistent failed queries.
+ */
+public class QueryMonitor implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(QueryMonitor.class);
+
+  private static final Random random = new Random();
+  private static final Ticker CURRENT_TIME_MILLIS_TICKER = new Ticker() {
+    @Override
+    public long read() {
+      return System.currentTimeMillis();
+    }
+  };
+
+  private static final long BASE_WAITING_TIME_MS = 10000;
+  private static final int SHUTDOWN_TIMEOUT_MS = 5000;
+
+  private final Ticker ticker;
+  private final long retryBackoffMaxMs;
+  private final KsqlEngine ksqlEngine;
+  private final ExecutorService executor;
+  private final Map<QueryId, RetryEvent> queriesRetries = new HashMap<>();
+
+  private volatile boolean closed = false;
+
+  public QueryMonitor(final KsqlConfig ksqlConfig, final KsqlEngine ksqlEngine) {
+    this(
+        ksqlConfig,
+        ksqlEngine,
+        Executors.newSingleThreadExecutor(r -> new Thread(r, "QueryMonitor")),
+        CURRENT_TIME_MILLIS_TICKER
+    );
+  }
+
+  @VisibleForTesting
+  QueryMonitor(
+      final KsqlConfig ksqlConfig,
+      final KsqlEngine ksqlEngine,
+      final ExecutorService executor,
+      final Ticker ticker
+  ) {
+    this.retryBackoffMaxMs = ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS);
+    this.ksqlEngine = ksqlEngine;
+    this.executor = executor;
+    this.ticker = ticker;
+  }
+
+  public void start() {
+    executor.execute(new Runner());
+    executor.shutdown();
+  }
+
+  @Override
+  public void close() {
+    closed = true;
+
+    try {
+      executor.awaitTermination(SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  boolean isClosed() {
+    return closed;
+  }
+
+  void restartFailedQueries() {
+    ksqlEngine.getPersistentQueries().stream()
+        .forEach(query -> {
+          final QueryId queryId = query.getQueryId();
+          final KafkaStreams.State queryState = query.getState();
+
+          // If the query was restarted previously, check if it needs another restart; or if the
+          // query is now up and running, then remove it from the restart list.
+          if (queriesRetries.containsKey(queryId)) {
+            final RetryEvent retryEvent = queriesRetries.get(queryId);
+            if (ticker.read() > retryEvent.nextRestartTimeMs()) {
+              if (queryState == KafkaStreams.State.ERROR) {
```

Comment:
Just wondering, we restart queries that are in NOT_RUNNING state initially. But here, we keep trying for queries that are in ERROR state. Why? How can a restart recover from an error?

Reply:
Currently, when a query goes into ERROR state, it stays there forever. There are errors like network latencies or misconfigured topics permissions, etc, that could cause this. In the case of network latencies, for instance, this retry mechanism will keep restarting the query until the network latency issue is gone, thus recovering a query from an error.

Other user errors such as misconfigured ACLs, can also be restarted until the issue is fixed by the user. The user will wait up to 2min after the fix to see the query running again (after the restart).